### PR TITLE
Testing Week

### PR DESCRIPTION
**Pull request by michael-sterling**
* BREAKING: [#11](https://github.com/Faithlife/FaithlifeTesting/issues/11)
  * Renamed `.Select([etc.])` to `.HasValue([etc.])`
  * Renamed `.Assert([etc.])` to `.IsTrue([etc.])`
  * Renamed `AssertEx.Builder<T>` to `Assertable<T>`
* Minor: [#18](https://github.com/Faithlife/FaithlifeTesting/issues/18) Add `.AssertResponse()` extension for `Faithlife.WebRequests.AutoWebServiceResponse`s.
* Minor: [#17](https://github.com/Faithlife/FaithlifeTesting/issues/17) Add `AssertEx.WaitUntil(Func<T>)` family of methods for retrying an action until it succeeds.
* Minor: [#10](https://github.com/Faithlife/FaithlifeTesting/issues/10) Allow `AssertEx.Value<T>(T value)` overload to specify an explicit string name.
* Patch: [#14](https://github.com/Faithlife/FaithlifeTesting/issues/14) Remove reference to specific NUnit version.
* Patch: [#13](https://github.com/Faithlife/FaithlifeTesting/issues/13) Improve output when AssertEx is used inside NUnit `Assert.Multiple` blocks.
* Patch: [#9](https://github.com/Faithlife/FaithlifeTesting/issues/9) Improve output when encountering indexers on captured lists.
* Patch: [#8](https://github.com/Faithlife/FaithlifeTesting/issues/8) Allow empty lists of context.
* Patch: Improve XML-doc comments

### CODE REVIEW

File: Directory.Build.props [hunk 1]
```diff
@@ -2,8 +2,8 @@
 
   <PropertyGroup>
     <VersionPrefix>1.3.0</VersionPrefix>
```
Comment (josephdangerstewart) [inline Directory.Build.props:4]: I believe this PR is a breaking change no? Not opposed! Just double checking

Comment (michael-sterling) [reply Directory.Build.props:4]: I *mostly* put this PR up to get the thing building locally *at all*. Marked this "Draft" to reflect that.

File: src/Faithlife.Testing/DebugValueExpressionVisitor.cs [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+			{
+				var isFirst = true;
+				foreach (var anded in andedExpressions)
+				{
+					var isException = false;
+					try
+					{
+						var value = Expression.Lambda<Func<bool>>(anded).Compile().Invoke();
+
+						// Skip and-ed values that are true. They are boring.
```
Comment (cndouglas) [inline src/Faithlife.Testing/DebugValueExpressionVisitor.cs:80]: 😀

**Review by cndouglas — COMMENTED** [review]
First round of feedback.

To fix the build check, I think the GitHub Actions workflow .NET version needs to be updated to .NET 5 (for C# 9 support). PR: #19

File: Faithlife.Testing.Libronix/Faithlife.Testing.WebRequests.csproj [hunk 3]
```diff
@@ -0,0 +1,17 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.1</TargetFramework>
```
[the remark this answers is not in the document]
Comment (michael-sterling) [reply Faithlife.Testing.Libronix/Faithlife.Testing.WebRequests.csproj:4]: Good catch! Thanks.

File: Faithlife.Testing.Libronix/Faithlife.Testing.WebRequests.csproj [hunk 4]
```diff
@@ -0,0 +1,17 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
[the remark this answers is not in the document]
Comment (michael-sterling) [reply Faithlife.Testing.Libronix/Faithlife.Testing.WebRequests.csproj:1]: This is intentional: I want it to be a separate NuGet package, so `Faithlife.Testing` itself doesn't have the extra baggage of additional dependencies, but I do *not* want callers to have the extra boilerplate of having an entirely new namespace to import once they *do* reference this package.

File: Directory.Build.props [hunk 5]
```diff
@@ -1,11 +1,11 @@
...
 
   <PropertyGroup>
-    <VersionPrefix>1.3.0</VersionPrefix>
+    <VersionPrefix>2.0.0</VersionPrefix>
     <LangVersion>9.0</LangVersion>
-    <Nullable>enable</Nullable>
+    <Nullable>disable</Nullable>
     <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
-    <NoWarn>$(NoWarn);1591;1998;NU5105;CA1014</NoWarn>
+    <NoWarn>$(NoWarn);1591;1998;NU5105;CA1014;CA1014;CA2007;CA1031</NoWarn>
```
Comment (cndouglas) [inline Directory.Build.props:8]: Nit: Duplicated rule
```suggestion
    <NoWarn>$(NoWarn);1591;1998;NU5105;CA1014;CA2007;CA1031</NoWarn>
```

File: Faithlife.Testing.Libronix/WebServiceResponseExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+			var properties = DtoInfo.GetInfo(typeof(TResponse))
+				.Properties
+				.Select(p => (p.Name, IsContent: IsContentProperty(p.Name), Value: p.GetValue(response)))
+				.Where(p => !IsDefault(p.Value))
+				.ToList();
+
+			var assertable = AssertEx.HasValue(response, "response")
+				.Context(GetContext(exception, properties));
+
+			if (properties.Any(p => p.Name == statusCodeString && p.Value.GetType().IsClass))
```
Comment (michael-sterling) [inline Faithlife.Testing.Libronix/WebServiceResponseExtensions.cs:52]: This should be 
```suggestion
			if (properties.Any(p => p.IsContent && p.Value.GetType().IsClass))
```

File: Faithlife.Testing.Libronix/Faithlife.Testing.WebRequests.csproj (hunk 4 above)
[the remark this answers is not in the document]
Comment (cndouglas) [reply Faithlife.Testing.Libronix/Faithlife.Testing.WebRequests.csproj:1]: I'm suggest renaming the _file directory_ (not the namespace) from `Faithlife.Testing.Libronix` to `Faithlife.Testing.WebRequests`. The directory is currently `Faithlife.Testing.Libronix/`, but the code is for `Faithlife.WebRequests`, not `Libronix`.

File: ReleaseNotes.md [hunk 7]
```diff
@@ -1,5 +1,20 @@
 # Release Notes
 
+## 2.0.0
+
+* BREAKING: [#11](https://github.com/Faithlife/FaithlifeTesting/issues/11)
+  * Renamed `.Select([etc.])` to `.HasValue([etc.])`
+  * Renamed `.Assert([etc.])` to `.IsTrue([etc.])`
+  * Renamed `AssertEx.Builder<T>` to `Assertible<T>`
```
Comment (cndouglas) [inline ReleaseNotes.md:8]: Nit: spelling
```suggestion
  * Renamed `AssertEx.Builder<T>` to `Assertable<T>`
```

File: ReleaseNotes.md [hunk 8]
```diff
@@ -1,5 +1,20 @@
...
+* Minor: [#18](https://github.com/Faithlife/FaithlifeTesting/issues/18) Add `.AssertResponse()` extension for `Faithlife.WebRequests.AutoWebServiceResponse`s.
+* Minor: [#17](https://github.com/Faithlife/FaithlifeTesting/issues/17) Add `AssertEx.WaitUntil(Func<T>)` family of methods for retrying an action until it succeeds.
+* Minor: [#10](https://github.com/Faithlife/FaithlifeTesting/issues/10) Allow `AssertEx.Value<T>(T value)` overload to specify an explicit string name.
+* Patch: [#14](https://github.com/Faithlife/FaithlifeTesting/issues/14) Remove reference to specific NUnit version.
+* Patch: [#13](https://github.com/Faithlife/FaithlifeTesting/issues/13) Improve output when AssertEx is used inside NUnit `Assert.Multiple` blocks.
+* Patch: [#9](https://github.com/Faithlife/FaithlifeTesting/issues/9) Imprive output when encountering indexers on captured lists.
```
Comment (cndouglas) [inline ReleaseNotes.md:14]: Nit: spelling
```suggestion
* Patch: [#9](https://github.com/Faithlife/FaithlifeTesting/issues/9) Improve output when encountering indexers on captured lists.
```

File: src/Faithlife.Testing/TestFrameworks/WaitUntilAssertable.cs [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+{
+	/// <summary>
+	/// Waits for a value which pases provided assertions.
+	/// </summary>
+	public sealed class WaitUntilAssertable<T>
+		where T : class
+	{
+		internal WaitUntilAssertable(Func<Task<Assertable<T>>> getAssertible, TimeSpan? timeout = null)
+		{
+			m_getAssertable = getAssertible;
```
Comment (cndouglas) [inline src/Faithlife.Testing/TestFrameworks/WaitUntilAssertable.cs:19]: Nit: spelling
```suggestion
		internal WaitUntilAssertable(Func<Task<Assertable<T>>> getAssertable, TimeSpan? timeout = null)
		{
			m_getAssertable = getAssertable;
```

File: src/Faithlife.Testing/TestFrameworks/AssertionFailedException.cs [hunk 10]
```diff
@@ -0,0 +1,27 @@
+using System;
+using System.Runtime.Serialization;
+
+namespace Faithlife.Testing.TestFrameworks
+{
+	/// <summary>
+	/// Represents the default exception in case no test framework is configured.
+	/// </summary>
+	[Serializable]
+
```
Comment (michael-sterling) [inline src/Faithlife.Testing/TestFrameworks/AssertionFailedException.cs:10]: Nit.
```suggestion
```

File: src/Faithlife.Testing/TestFrameworks/AssertionFailedException.cs [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+#pragma warning disable CA1032, RCS1194 // AssertionFailedException should never be constructed with an empty message
+	public class AssertionFailedException : Exception
```
Comment (michael-sterling) [inline src/Faithlife.Testing/TestFrameworks/AssertionFailedException.cs:12]: Nit.
```suggestion
	public sealed class AssertionFailedException : Exception
```

File: Faithlife.Testing.Libronix/Faithlife.Testing.WebRequests.csproj (hunk 4 above)
[the remark this answers is not in the document]
Comment (michael-sterling) [reply Faithlife.Testing.Libronix/Faithlife.Testing.WebRequests.csproj:1]: Ah, good catch, thanks!

[the remark this answers is not in the document]
Comment (michael-sterling) [reply Faithlife.Testing.Libronix/Faithlife.Testing.WebRequests.csproj]: 0f7cbf4a38e470a66791310bd92cb92ae3243507

File: Directory.Build.props [hunk 12]
```diff
@@ -1,11 +1,11 @@
 <Project>
 
   <PropertyGroup>
-    <VersionPrefix>1.3.0</VersionPrefix>
+    <VersionPrefix>2.0.0</VersionPrefix>
     <LangVersion>9.0</LangVersion>
-    <Nullable>enable</Nullable>
+    <Nullable>disable</Nullable>
```
Comment (StephenCleary) [inline Directory.Build.props:6]: Can we keep null checking?

Comment (michael-sterling) [reply Directory.Build.props:6]: I would love to, but I spent ~10 minutes trying to add the proper annotations and (a) I didn't feel like I was adding value, and (b) I did not have any end in sight. This code makes *heavy* and intentional use of `null` values, and slathering a `?` on everything didn't feel worth the delay in merging this PR.

I've created a task to turn it back on after this PR merges if you'd like to take a crack at it: https://github.com/Faithlife/FaithlifeTesting/issues/24

**Pull request merged**